src/en/developer-testing.md

```diff
@@ -0,0 +1,160 @@
+
+The intention of charm tests is to assert that the charm works well on the
+intended platform and performs the expected configuration steps. Examples of
+things to test in each charm is:
+
+- After install, expose, and adding of required relations, the service is
+  running correctly (such as listening on the intended ports).
+- Adding, removing, and re-adding a relation should work without error.
+- Setting configuration values should result in a change reflected in the
+  service's operation or configuration.
+
+## Where to put tests
+
+The charm directory should contain a sub-directory named 'tests'.  This
+directory will be scanned by a test runner for executable files. The executable
+files will be run in lexical order by the test runner, with a juju environment.
+The tests can make the following assumptions:
+
+- A minimal install of the release of Ubuntu which the charm is targeted at
+  will be available.
+- A version of juju is installed and available in the system path.
+- A juju environment with no services deployed inside it is already
+  bootstrapped, and will be the default for command line usage.
+- The CWD is the `tests` directory off the charm root.
+- Full access to other public charms will be available to build a solution of
+  your choice.
+- Tests should be self contained, meaning include or install the packages that
+  it needs to test the software.
+- Tests should run automatically and not require (such as passwords) or human
+  intervention to get a successful test result.
+
+If a tool is needed to perform a test and is not available in the Ubuntu
+archive, it can also be included in the `tests/` directory, as long as the file
+which contains it is not executable. Note that build tools cannot be assumed to
+be available on the testing system.
+
+## Test automation
+
+The test runner will be run automatically so all charm tests must be self
+contained, meaning tests must install or package the files required to test
+the charm.  The test runner will find and execute each test within that
+directory and produce a report.
+
+If tests exit with services still in the environment, the test runner may clean
+them up, whether by destroying the environment or destroying the services
+explicitly, and the machines may be terminated as well. Any artifacts needed
+from the test machines should be retrieved and displayed before the test exits.
+
+### Exit codes
+
+Upon exit, the test's exit code will be evaluated to mean the following:
+
+- 0: Test passed
+- 1: Failed test
+- 100: Test is skipped because of incomplete environment
+
+## charm proof
+
+The `charm-tools` package contains a static charm analysis tool called
+`charm proof`.  This tool checks the charm structure and gives Informational,
+Warning, and Error messages on potential issues with the charm structure. All
+charms should pass `charm proof` with Information messages only.
+Warning or Error messages indicate a problem that should be fixed in the charm.
```

Comment:
Would it help if I linked to the best practices or charm store policy documents?